### PR TITLE
Refactor component package

### PR DESCRIPTION
**Pull request by mrchtr**
Refactor component package as part of #643

### CODE REVIEW

File: src/fondant/component/data_io.py [hunk 1]
```diff
@@ -82,35 +82,7 @@ def partition_loaded_dataframe(self, dataframe: dd.DataFrame) -> dd.DataFrame:
...
-
-        # add subset prefix to columns
-        subset_df = subset_df.rename(
-            columns={col: subset_name + "_" + col for col in subset_df.columns},
-        )
-
-        return subset_df
-
+    # TODO: probably not needed anymore!
     def _load_index(self) -> dd.DataFrame:
```
Comment (RobbeSneyders) [inline src/fondant/component/data_io.py:86]: This can be removed.

File: src/fondant/component/data_io.py [hunk 2]
```diff
@@ -135,20 +108,34 @@ def load_dataframe(self) -> dd.DataFrame:
...
-            # left joins -> filter on index
-            dataframe = dd.merge(
-                dataframe,
-                subset_df,
-                left_index=True,
-                right_index=True,
-                how="left",
+        dataframe = None
+        field_mapping = self.manifest.field_mapping
+        for location, fields in field_mapping.items():
```
Comment (RobbeSneyders) [inline src/fondant/component/data_io.py:113]: We need to make sure that we read the location containing the latest index first, so the left join below is correct.

File: src/fondant/component/data_io.py [hunk 3]
```diff
@@ -135,20 +108,34 @@ def load_dataframe(self) -> dd.DataFrame:
...
+                index="id",
+                calculate_divisions=True,
             )
 
+            if dataframe is None:
+                # ensure that the index is set correctly and divisions are known.
+                dataframe = partial_df
+            else:
+                dask_divisions = partial_df.set_index("id").divisions
+                unique_divisions = list(dict.fromkeys(list(dask_divisions)))
```
Comment (RobbeSneyders) [inline src/fondant/component/data_io.py:126]: Why is this needed? And why are you using `dict.fromkeys`? I would assume we can use `set`?

File: src/fondant/component/data_io.py [hunk 4]
```diff
@@ -135,20 +108,34 @@ def load_dataframe(self) -> dd.DataFrame:
...
+                columns=fields,
+                index="id",
+                calculate_divisions=True,
             )
 
+            if dataframe is None:
+                # ensure that the index is set correctly and divisions are known.
+                dataframe = partial_df
+            else:
+                dask_divisions = partial_df.set_index("id").divisions
```
Comment (RobbeSneyders) [inline src/fondant/component/data_io.py:125]: I would expect us to use the divisions from the already loaded dataframe, not from the newly loaded one.

File: src/fondant/component/data_io.py [hunk 5]
```diff
@@ -170,79 +157,46 @@ def write_dataframe(
         dataframe: dd.DataFrame,
         dask_client: t.Optional[Client] = None,
     ) -> None:
-        write_tasks = []
+        columns_to_produce = [
+            column_name for column_name, field in self.component_spec.produces.items()
+        ]
 
-        dataframe.index = dataframe.index.rename("id")
```
Comment (RobbeSneyders) [inline src/fondant/component/data_io.py:175]: We are assuming that the index is called "id" in the loader, so I think we still need this?

File: src/fondant/component/executor.py [hunk 6]
```diff
@@ -569,12 +549,16 @@ def _execute_component(
 
         # Clear divisions if component spec indicates that the index is changed
         if self._infer_index_change():
+            # TODO: might causing issues for merging components
```
Comment (RobbeSneyders) [inline src/fondant/component/executor.py:552]: We still need divisions, but we need to make sure to reset and recalculate them if the index changes. See the [Dask documentation](https://docs.dask.org/en/stable/generated/dask.dataframe.DataFrame.map_partitions.html#:~:text=Also%20note%20that%20the%20index%20and%20divisions%20are%20assumed%20to%20remain%20unchanged.%20If%20the%20function%20you%E2%80%99re%20mapping%20changes%20the%20index/divisions%2C%20you%E2%80%99ll%20need%20to%20clear%20them%20afterwards%3A).

**Review by RobbeSneyders — COMMENTED** [review]
Just a quick partial review. I'll need to take my time to dive a bit deeper to review with Dask performance in mind.

File: src/fondant/component/data_io.py (hunk 2 above)
Comment (mrchtr) [reply src/fondant/component/data_io.py:97]: I've updated the `field_mapping` method. The method returns an immutable OrderedDict now. The first entry is the location of the dataset of the latest index.

File: src/fondant/component/data_io.py (hunk 4 above)
Comment (mrchtr) [reply src/fondant/component/data_io.py]: To summarise theory to double check that I didn't missed a detail:
- we are saving parquet files with index and devision information
- start reading the parquet file with the most recent index (index.name = id)
- loading iterative the next parquet file (containing the same divisions and index.name = id) and apply a left join

If so, we can trust the division and index information of the parquet files and don't need to set it here.

File: src/fondant/component/data_io.py (hunk 2 above)
Comment (PhilippeMoussalli) [reply src/fondant/component/data_io.py:97]: I might have missed something but it seems to be that we're currently loading/merging all the fields defined in the manifest since that's what `field_mapping` contains. Shouldn't we only base ourselves on the fields defined in the `consumes` section of the component spec?

Comment (RobbeSneyders) [reply src/fondant/component/data_io.py:97]: Good catch :sweat_smile:

I think this should work:
```python
fields = [field for field in fields if 
          (field in self.component_spec.consumes or field == DEFAULT_INDEX_NAME)]
```

File: src/fondant/component/data_io.py (hunk 4 above)
Comment (RobbeSneyders) [reply src/fondant/component/data_io.py]: Yes, that's my understanding as well.

File: src/fondant/component/data_io.py (hunk 2 above)
Comment (mrchtr) [reply src/fondant/component/data_io.py:97]: I've started the implementation of the consumes and produces in #665.
Maybe it is easier to include the changes of the component package there here as well! I'll take a look. 

Basically, the idea was the following:

- Load the dataframe based on the manifest, later on renaming the columns based on the defined consumes before the dataset is passed to the transform method. 
- When produce is defined renaming the columns of the output dataframe accordingly to the produce mapping, write it to the bucket and use the written column names for the output manifest. 

Am I overlooking something?

Comment (PhilippeMoussalli) [reply src/fondant/component/data_io.py:99]: > Good catch 😅
> 
> I think this should work:
> 
> ```python
> fields = [field for field in fields if 
>           (field in self.component_spec.consumes or field == DEFAULT_INDEX_NAME)]
> ```

I don't see why we need the manifest field mapping in the first place. 

Can't we do something like this (a bit similar to how it was before):

```
for field_name in self.component_spec.consumes.items():
             field_location = self.manifest.fields[field_name].location
```

**Comment by RobbeSneyders** [comment]
I merged the base branch into this one so we can review @mrchtr. Please double check that I did it correctly :)

File: src/fondant/component/component.py [hunk 7]
```diff
@@ -16,7 +16,9 @@ class BaseComponent:
         **kwargs: The provided user arguments are passed in as keyword arguments
     """
 
-    def __init__(self, spec: ComponentSpec, **kwargs):
+    def __init__(
+        self, spec: ComponentSpec, schema: t.Optional[t.Dict[str, str]] = None, **kwargs
```
Comment (RobbeSneyders) [inline src/fondant/component/component.py:20]: Where is this used? If we do need it, can you add it to the docstring?

File: src/fondant/component/data_io.py [hunk 8]
```diff
@@ -25,9 +27,11 @@ def __init__(
         manifest: Manifest,
         component_spec: ComponentSpec,
         input_partition_rows: t.Optional[int] = None,
+        consumes: t.Optional[t.Dict[str, str]] = None,
```
Comment (RobbeSneyders) [inline src/fondant/component/data_io.py:30]: I assume these are changes related to the pipeline interface change? Can we keep those out of this PR for now?

File: src/fondant/component/data_io.py [hunk 9]
```diff
@@ -82,77 +86,58 @@ def partition_loaded_dataframe(self, dataframe: dd.DataFrame) -> dd.DataFrame:
...
         dataframe = self.partition_loaded_dataframe(dataframe)
 
         logging.info(f"Columns of dataframe: {list(dataframe.columns)}")
 
+        # rename columns accordingly to the consumes
+        if self.consumes:
+            reverted_consumes = {v: k for k, v in self.consumes.items()}
+            dataframe = dataframe.rename(columns=reverted_consumes)
+
+        logging.info(f"Columns of components dataframe: {list(dataframe.columns)}")
```
Comment (RobbeSneyders) [inline src/fondant/component/data_io.py:140]: Same here, can we keep these changes out for now?

File: src/fondant/component/data_io.py [hunk 10]
```diff
@@ -162,87 +147,64 @@ def __init__(
         *,
         manifest: Manifest,
         component_spec: ComponentSpec,
+        produces: t.Optional[t.Dict[str, str]] = None,
```
Comment (RobbeSneyders) [inline src/fondant/component/data_io.py:150]: Same here, can we keep these changes out for now?

File: src/fondant/component/data_io.py [hunk 11]
```diff
@@ -162,87 +147,64 @@ def __init__(
...
-                subset_name=subset_name,
-                subset_spec=subset_spec,
-            )
-            write_tasks.append(write_subset_task)
+        dataframe = dataframe[columns_to_produce]
+
+        # Rename produces accordingly
+        if self.produces:
+            reverted_produces = {v: k for k, v in self.produces.items()}
+            dataframe = dataframe.rename(columns=reverted_produces)
```
Comment (RobbeSneyders) [inline src/fondant/component/data_io.py:174]: Same here, can we keep these changes out for now?

File: src/fondant/component/executor.py
Comment (RobbeSneyders) [inline src/fondant/component/executor.py:1]: I think most changes in this file are related to the changed interface. Can we keep this out of this PR for now?

**Review by RobbeSneyders — COMMENTED** [review]
Can we keep any changes related to the changed interface out of this PR and in #665? It's a bit hard to keep track otherwise.

**Review by RobbeSneyders — APPROVED** [review]
Thanks @mrchtr! Good to go!
